### PR TITLE
github-ci: add fedora-34, delete fedora-28,29

### PR DESCRIPTION
**Pull request by VitaliyaIoffe**
The patch includes 4 commits.

1) First fixes uninitialized variables in the build.
Build for Fedora 34 is breaking out due to uninitialized variables in
a few places:
For example,
```
[100%] Built target merger.test
/source/build/usr/src/debug/tarantool-2.9.0.116/src/box/sql.c: In function 'tarantoolSqlNextSeqId':
/source/build/usr/src/debug/tarantool-2.9.0.116/src/box/sql.c:1186:13: error: 'key' may be used uninitialized [-Werror=maybe-uninitialized]
1186 |         if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
```

Needed for: #6074

2) Second provides a new workflow for fedora-34

Part of: #6074

3) Next erases old fedora's releases from workflows, cause they are not supported for more than 1.5 years.

Part of: #6074

4) Finally, make able to save packages in S3 buckets.
Closes: #6074

### CODE REVIEW

File: .github/workflows/fedora_34.yml [hunk 1]
```diff
@@ -42,7 +42,8 @@ jobs:
           GPG_SECRET_KEY: ${{ secrets.GPG_SECRET_KEY }}
           GPG_SIGN_KEY: ${{ secrets.GPG_SIGN_KEY }}
           OS: 'fedora'
-          DIST: '28'
+          DIST: '34'
+          DOCKER_REPO: 'vitaliiaioffe/tarantool'
```
Comment (VitaliyaIoffe) [inline .github/workflows/fedora_34.yml:46]: pls, fix

File: tools/update_repo.sh [hunk 2]
```diff
@@ -30,7 +30,7 @@ function get_os_dists {
     elif [ "$os" == "el" ]; then
         alldists='6 7 8'
     elif [ "$os" == "fedora" ]; then
-        alldists='27 28 29 30 31 32 33'
+        alldists='27 28 29 30 31 32 33 34'
```
Comment (VitaliyaIoffe) [inline tools/update_repo.sh:33]: @Totktonada Should I delete previous versions here?

Comment (Totktonada) [reply tools/update_repo.sh:33]: I think it can be useful if we'll need to manually push something to an old repository. So would keep old distros.

Comment (Totktonada) [reply tools/update_repo.sh:33]: Nit: I would move the update_repo commit before adding the workflow: this way each commit would be fully working.

Comment (VitaliyaIoffe) [reply tools/update_repo.sh:33]: update_repo commit was moved

**Comment by Totktonada** [comment]
Regarding description:

> build: fix uninitialized variables 

It looks like you fixed a bug rather than suppress a false-positive warning.

Regarding extra initialization in `tarantoolSqlNextSeqId()`:

I built tarantool with LTO using GCC 11:

```sh
CC=gcc-11.1.0 CXX=g++-11.1.0 cmake . -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_BACKTRACE=ON -DENABLE_DIST=ON -DENABLE_FEEDBACK_DAEMON=OFF -DENABLE_BUNDLED_LIBCURL=OFF -DENABLE_LTO=ON && make -j
```

Disassembled the function code (`objdump -d src/tarantool`, cut off the function):

<details>
<summary>tarantoolSqlNextSeqId</summary>

```asm
00000000000c5690 <tarantoolSqlNextSeqId>:
   c5690:	55                   	push   %rbp
   c5691:	31 f6                	xor    %esi,%esi
   c5693:	48 89 e5             	mov    %rsp,%rbp
   c5696:	41 54                	push   %r12
   c5698:	53                   	push   %rbx
   c5699:	48 8d 55 e0          	lea    -0x20(%rbp),%rdx
   c569d:	48 89 fb             	mov    %rdi,%rbx
   c56a0:	48 8d 4d e1          	lea    -0x1f(%rbp),%rcx
   c56a4:	4c 8d 45 d8          	lea    -0x28(%rbp),%r8
   c56a8:	bf 1c 01 00 00       	mov    $0x11c,%edi
   c56ad:	48 83 ec 20          	sub    $0x20,%rsp
   c56b1:	c6 45 e0 90          	movb   $0x90,-0x20(%rbp)
   c56b5:	e8 36 fd fa ff       	callq  753f0 <box_index_max>
   c56ba:	85 c0                	test   %eax,%eax
   c56bc:	0f 85 8e 00 00 00    	jne    c5750 <tarantoolSqlNextSeqId+0xc0>
   c56c2:	48 8b 7d d8          	mov    -0x28(%rbp),%rdi
   c56c6:	41 89 c4             	mov    %eax,%r12d
   c56c9:	48 85 ff             	test   %rdi,%rdi
   c56cc:	74 52                	je     c5720 <tarantoolSqlNextSeqId+0x90>
   c56ce:	31 f6                	xor    %esi,%esi
   c56d0:	ba 01 00 00 00       	mov    $0x1,%edx
   c56d5:	e8 a6 fa ff ff       	callq  c5180 <tuple_field_with_type.lto_priv.1.lto_priv.0>
   c56da:	48 85 c0             	test   %rax,%rax
   c56dd:	74 71                	je     c5750 <tarantoolSqlNextSeqId+0xc0>
   c56df:	0f b6 10             	movzbl (%rax),%edx
   c56e2:	80 fa ce             	cmp    $0xce,%dl
   c56e5:	74 59                	je     c5740 <tarantoolSqlNextSeqId+0xb0>
   c56e7:	77 27                	ja     c5710 <tarantoolSqlNextSeqId+0x80>
   c56e9:	80 fa cc             	cmp    $0xcc,%dl
   c56ec:	74 4a                	je     c5738 <tarantoolSqlNextSeqId+0xa8>
   c56ee:	80 fa cd             	cmp    $0xcd,%dl
   c56f1:	75 0d                	jne    c5700 <tarantoolSqlNextSeqId+0x70>
   c56f3:	0f b7 50 01          	movzwl 0x1(%rax),%edx
   c56f7:	66 c1 c2 08          	rol    $0x8,%dx
   c56fb:	0f b7 d2             	movzwl %dx,%edx
   c56fe:	66 90                	xchg   %ax,%ax
   c5700:	48 89 13             	mov    %rdx,(%rbx)
   c5703:	48 83 c4 20          	add    $0x20,%rsp
   c5707:	44 89 e0             	mov    %r12d,%eax
   c570a:	5b                   	pop    %rbx
   c570b:	41 5c                	pop    %r12
   c570d:	5d                   	pop    %rbp
   c570e:	c3                   	retq   
   c570f:	90                   	nop
   c5710:	80 fa cf             	cmp    $0xcf,%dl
   c5713:	75 eb                	jne    c5700 <tarantoolSqlNextSeqId+0x70>
   c5715:	48 8b 50 01          	mov    0x1(%rax),%rdx
   c5719:	48 0f ca             	bswap  %rdx
   c571c:	eb e2                	jmp    c5700 <tarantoolSqlNextSeqId+0x70>
   c571e:	66 90                	xchg   %ax,%ax
   c5720:	48 c7 03 00 00 00 00 	movq   $0x0,(%rbx)
   c5727:	48 83 c4 20          	add    $0x20,%rsp
   c572b:	44 89 e0             	mov    %r12d,%eax
   c572e:	5b                   	pop    %rbx
   c572f:	41 5c                	pop    %r12
   c5731:	5d                   	pop    %rbp
   c5732:	c3                   	retq   
   c5733:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)
   c5738:	0f b6 50 01          	movzbl 0x1(%rax),%edx
   c573c:	eb c2                	jmp    c5700 <tarantoolSqlNextSeqId+0x70>
   c573e:	66 90                	xchg   %ax,%ax
   c5740:	8b 50 01             	mov    0x1(%rax),%edx
   c5743:	0f ca                	bswap  %edx
   c5745:	89 d2                	mov    %edx,%edx
   c5747:	eb b7                	jmp    c5700 <tarantoolSqlNextSeqId+0x70>
   c5749:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
   c5750:	41 bc ff ff ff ff    	mov    $0xffffffff,%r12d
   c5756:	eb ab                	jmp    c5703 <tarantoolSqlNextSeqId+0x73>
   c5758:	0f 1f 84 00 00 00 00 	nopl   0x0(%rax,%rax,1)
   c575f:	00 
```
</details>

Applied the patch:

```diff
diff --git a/src/box/sql.c b/src/box/sql.c
index 790ca7f70..5c4daf208 100644
--- a/src/box/sql.c
+++ b/src/box/sql.c
@@ -1180,7 +1180,7 @@ sql_debug_info(struct info_handler *h)
 int
 tarantoolSqlNextSeqId(uint64_t *max_id)
 {
-       char key[16];
+       char key[16] = {'\0'};
        struct tuple *tuple;
        char *key_end = mp_encode_array(key, 0);
        if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
```

Built and disassembled it again:

<details>
<summary>tarantoolSqlNextSeqId</summary>

```asm
00000000000c5570 <tarantoolSqlNextSeqId>:
   c5570:	55                   	push   %rbp
   c5571:	66 0f ef c0          	pxor   %xmm0,%xmm0  !!
   c5575:	31 f6                	xor    %esi,%esi
   c5577:	48 89 e5             	mov    %rsp,%rbp
   c557a:	41 54                	push   %r12
   c557c:	53                   	push   %rbx
   c557d:	48 8d 55 e0          	lea    -0x20(%rbp),%rdx
   c5581:	48 89 fb             	mov    %rdi,%rbx
   c5584:	48 8d 4d e1          	lea    -0x1f(%rbp),%rcx
   c5588:	4c 8d 45 d8          	lea    -0x28(%rbp),%r8
   c558c:	bf 1c 01 00 00       	mov    $0x11c,%edi
   c5591:	48 83 ec 20          	sub    $0x20,%rsp
   c5595:	0f 29 45 e0          	movaps %xmm0,-0x20(%rbp)  !!
   c5599:	c6 45 e0 90          	movb   $0x90,-0x20(%rbp)
   c559d:	e8 4e fe fa ff       	callq  753f0 <box_index_max>
   c55a2:	85 c0                	test   %eax,%eax
   c55a4:	0f 85 96 00 00 00    	jne    c5640 <tarantoolSqlNextSeqId+0xd0>
   c55aa:	48 8b 7d d8          	mov    -0x28(%rbp),%rdi
   c55ae:	41 89 c4             	mov    %eax,%r12d
   c55b1:	48 85 ff             	test   %rdi,%rdi
   c55b4:	74 5a                	je     c5610 <tarantoolSqlNextSeqId+0xa0>
   c55b6:	31 f6                	xor    %esi,%esi
   c55b8:	ba 01 00 00 00       	mov    $0x1,%edx
   c55bd:	e8 be fb ff ff       	callq  c5180 <tuple_field_with_type.lto_priv.1.lto_priv.0>
   c55c2:	48 85 c0             	test   %rax,%rax
   c55c5:	74 79                	je     c5640 <tarantoolSqlNextSeqId+0xd0>
   c55c7:	0f b6 10             	movzbl (%rax),%edx
   c55ca:	80 fa ce             	cmp    $0xce,%dl
   c55cd:	74 61                	je     c5630 <tarantoolSqlNextSeqId+0xc0>
   c55cf:	77 2f                	ja     c5600 <tarantoolSqlNextSeqId+0x90>
   c55d1:	80 fa cc             	cmp    $0xcc,%dl
   c55d4:	74 52                	je     c5628 <tarantoolSqlNextSeqId+0xb8>
   c55d6:	80 fa cd             	cmp    $0xcd,%dl
   c55d9:	75 15                	jne    c55f0 <tarantoolSqlNextSeqId+0x80>
   c55db:	0f b7 50 01          	movzwl 0x1(%rax),%edx
   c55df:	66 c1 c2 08          	rol    $0x8,%dx
   c55e3:	0f b7 d2             	movzwl %dx,%edx
   c55e6:	66 2e 0f 1f 84 00 00 	nopw   %cs:0x0(%rax,%rax,1)
   c55ed:	00 00 00 
   c55f0:	48 89 13             	mov    %rdx,(%rbx)
   c55f3:	48 83 c4 20          	add    $0x20,%rsp
   c55f7:	44 89 e0             	mov    %r12d,%eax
   c55fa:	5b                   	pop    %rbx
   c55fb:	41 5c                	pop    %r12
   c55fd:	5d                   	pop    %rbp
   c55fe:	c3                   	retq   
   c55ff:	90                   	nop
   c5600:	80 fa cf             	cmp    $0xcf,%dl
   c5603:	75 eb                	jne    c55f0 <tarantoolSqlNextSeqId+0x80>
   c5605:	48 8b 50 01          	mov    0x1(%rax),%rdx
   c5609:	48 0f ca             	bswap  %rdx
   c560c:	eb e2                	jmp    c55f0 <tarantoolSqlNextSeqId+0x80>
   c560e:	66 90                	xchg   %ax,%ax
   c5610:	48 c7 03 00 00 00 00 	movq   $0x0,(%rbx)
   c5617:	48 83 c4 20          	add    $0x20,%rsp
   c561b:	44 89 e0             	mov    %r12d,%eax
   c561e:	5b                   	pop    %rbx
   c561f:	41 5c                	pop    %r12
   c5621:	5d                   	pop    %rbp
   c5622:	c3                   	retq   
   c5623:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)
   c5628:	0f b6 50 01          	movzbl 0x1(%rax),%edx
   c562c:	eb c2                	jmp    c55f0 <tarantoolSqlNextSeqId+0x80>
   c562e:	66 90                	xchg   %ax,%ax
   c5630:	8b 50 01             	mov    0x1(%rax),%edx
   c5633:	0f ca                	bswap  %edx
   c5635:	89 d2                	mov    %edx,%edx
   c5637:	eb b7                	jmp    c55f0 <tarantoolSqlNextSeqId+0x80>
   c5639:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
   c5640:	41 bc ff ff ff ff    	mov    $0xffffffff,%r12d
   c5646:	eb ab                	jmp    c55f3 <tarantoolSqlNextSeqId+0x83>
   c5648:	0f 1f 84 00 00 00 00 	nopl   0x0(%rax,%rax,1)
   c564f:	00 
```
</details>

Consider two exclamation marks: we actually perform this zeroing during execution of the function. If this code is on a hot path, it is undesirable. I would ask someone who work on SQL to look at this.

The nice thing is that `mp_encode_array(key, 0);` actually becomes just move of `0x90`.

What is also interesting: it seems, GCC 11 actually complains about `key_end` ("by argument 4 <...> to ‘box_index_max’"). However if we'll make `key` being array of size 1 (we only need to store `0x90`), GCC will be okay. It seems, the compiler complain about passing a pointer to uninitialized memory, which we explicitly allocate, and not about memory, which is known to be out of array bounds.

So there are several ways to eliminate the warning and don't add extra zeroing code.

```diff
diff --git a/src/box/sql.c b/src/box/sql.c
index 790ca7f70..41732f142 100644
--- a/src/box/sql.c
+++ b/src/box/sql.c
@@ -1180,7 +1180,7 @@ sql_debug_info(struct info_handler *h)
 int
 tarantoolSqlNextSeqId(uint64_t *max_id)
 {
-	char key[16];
+	char key[1];
 	struct tuple *tuple;
 	char *key_end = mp_encode_array(key, 0);
 	if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
```

```diff
diff --git a/src/box/sql.c b/src/box/sql.c
index 790ca7f70..51dfefa0c 100644
--- a/src/box/sql.c
+++ b/src/box/sql.c
@@ -1180,10 +1180,10 @@ sql_debug_info(struct info_handler *h)
 int
 tarantoolSqlNextSeqId(uint64_t *max_id)
 {
-	char key[16];
+	char key;
 	struct tuple *tuple;
-	char *key_end = mp_encode_array(key, 0);
-	if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
+	char *key_end = mp_encode_array(&key, 0);
+	if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, &key,
 			  key_end, &tuple) != 0)
 		return -1;
```

```diff
diff --git a/src/box/sql.c b/src/box/sql.c
index 790ca7f70..1a17129aa 100644
--- a/src/box/sql.c
+++ b/src/box/sql.c
@@ -1180,9 +1180,10 @@ sql_debug_info(struct info_handler *h)
 int
 tarantoolSqlNextSeqId(uint64_t *max_id)
 {
-	char key[16];
+	/* The key is array of size zero. */
+	const char *key = "\x90";
+	const char *key_end = key + 1;
 	struct tuple *tuple;
-	char *key_end = mp_encode_array(key, 0);
 	if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
 			  key_end, &tuple) != 0)
 		return -1;
```

I would ask someone who work on SQL, what is better to do.

Regarding extra initialization in tests: I think it is okay.

**Comment by Totktonada** [comment]
The rest looks okay, so I'll approve it.

**Review by Totktonada — APPROVED** [review]

**Comment by kyukhin** [comment]
Here's a minimalistic fix.
I guess we should (more or less) care about extra inits (which might debgade if on hot path) in code itself and not care at all while in tests.
``` C
--- a/src/box/sql.c
+++ b/src/box/sql.c        
@@ -1180,9 +1180,10 @@ sql_debug_info(struct info_handler *h)
 int                                             
 tarantoolSqlNextSeqId(uint64_t *max_id)
 {               
-  char key[16];       
+       char key[1];
        struct tuple *tuple;
        char *key_end = mp_encode_array(key, 0);
+       assert(key_end - key == 1);
        if (box_index_max(BOX_SEQUENCE_ID, 0 /* PK */, key,
                          key_end, &tuple) != 0)
                return -1;
diff --git a/test/box/function1.c b/test/box/function1.c
index 84bbdd7..eb2d1e2 100644
--- a/test/box/function1.c
+++ b/test/box/function1.c
@@ -259,6 +259,7 @@ test_return_mp(box_function_ctx_t *ctx, const char *args, const char *args_end)
        (void) args;
        (void) args_end;
        char buf[512];
+       memset(buf, '\0', 512);
        char *pos = mp_encode_uint(buf, 1);
        int rc = box_return_mp(ctx, buf, pos);
        if (rc != 0)
diff --git a/test/sql-tap/gh-6024-funcs-return-bin.c b/test/sql-tap/gh-6024-funcs-return-bin.c
index ed90700..553be0f 100644
--- a/test/sql-tap/gh-6024-funcs-return-bin.c
+++ b/test/sql-tap/gh-6024-funcs-return-bin.c
@@ -14,6 +14,7 @@ ret_bin(box_function_ctx_t *ctx, const char *args, const char *args_end)
        (void)args_end;
        const char bin[] = "some varbinary string"; 
        char res[BUF_SIZE];
+       memset(res, '\0', BUF_SIZE);
        char *end = mp_encode_bin(res, bin, sizeof(bin));
        box_return_mp(ctx, res, end);
        return 0;
@@ -27,6 +28,7 @@ ret_uuid(box_function_ctx_t *ctx, const char *args, const char *args_end)
        struct tt_uuid uuid;
        memset(&uuid, 0x11, sizeof(uuid));
        char res[BUF_SIZE];
+       memset(res, '\0', BUF_SIZE);
        char *end = mp_encode_uuid(res, &uuid);
        box_return_mp(ctx, res, end);
        return 0;
@@ -40,6 +42,7 @@ ret_decimal(box_function_ctx_t *ctx, const char *args, const char *args_end)
        decimal_t dec;
        decimal_from_string(&dec, "9999999999999999999.9999999999999999999");
        char res[BUF_SIZE];
+       memset(res, '\0', BUF_SIZE);
        char *end = mp_encode_decimal(res, &dec);
        box_return_mp(ctx, res, end);
        return 0;
diff --git a/test/sql-tap/sql_uuid.c b/test/sql-tap/sql_uuid.c
index 4032bba..93f1e42 100644
--- a/test/sql-tap/sql_uuid.c
+++ b/test/sql-tap/sql_uuid.c
@@ -27,6 +27,7 @@ is_uuid(box_function_ctx_t *ctx, const char *args, const char *args_end)
        }
 
        char res[BUF_SIZE];
+       memset(res, '\0', BUF_SIZE);
        char *end = mp_encode_bool(res, is_uuid);
        box_return_mp(ctx, res, end);
        return 0;
```
```

**Comment by Totktonada** [comment]
I forgot to remind about changelogs.

**Comment by VitaliyaIoffe** [comment]
> I forgot to remind about changelogs.

I added it to the last commit.

**Review by kyukhin — APPROVED** [review]

**Pull request merged**